File: server/src/utils/totp.js
const { authenticator } = require('otplib');

/**
 * Generates the current 6-digit TOTP code from the secret a user obtained
 * once from smartapi.angelbroking.com/enable-totp.
 * This is what lets the bot log in without a human typing a fresh code every time.
 */
function generateTotp(totpSecret) {
  if (!totpSecret) {
    throw new Error('TOTP secret is required to generate a login code');
  }
  // Angel One issues standard base32 TOTP secrets (30s step, 6 digits, SHA1) -
  // the otplib defaults already match that, so no extra options are needed.
  return authenticator.generate(totpSecret.replace(/\s+/g, ''));
}

module.exports = { generateTotp };

File: server/utils/totp.py
"""
TOTP (Time-based One-Time Password) generator for Angel One authentication
"""
import pyotp


def generate_totp(totp_secret: str) -> str:
    """
    Generates the current 6-digit TOTP code from the secret.
    
    Args:
        totp_secret: The TOTP secret obtained from Angel One
        
    Returns:
        6-digit TOTP code as string
        
    Raises:
        ValueError: If TOTP secret is not provided
    """
    if not totp_secret:
        raise ValueError("TOTP secret is required to generate a login code")
    
    # Remove any whitespace from the secret
    clean_secret = totp_secret.replace(" ", "")
    
    # Angel One uses standard TOTP (30s step, 6 digits, SHA1)
    totp = pyotp.TOTP(clean_secret)
    return totp.now()

# Made with Bob
